kornia/losses/total_variation.py

```diff
@@ -0,0 +1,44 @@
+import torch
+import torch.nn as nn
+
+
+class TotalVariation(nn.Module):
+    r"""Computes the Total Variation according to
+    [1] https://en.wikipedia.org/wiki/Total_variation
+    Shape:
+        - Input: :math:`(N, C, H, W)` or :math:`(C, H, W)` where C = number of classes.
+        - Target: :math:`(N,)` or :math:`()`
+    Examples:
+        >>> kornia.losses.total_variation(torch.ones(3,4,4)) # tensor(0.)
+        >>> tv = kornia.losses.TotalVariation()
+        >>> output = tv(torch.ones(2,3,4,4)) # tensor([0., 0.])
+        >>> output.backward()
+    """
+    def __init__(self) -> None:
+        super(TotalVariation, self).__init__()
+
+    def forward(  # type: ignore
+            self, img) -> torch.Tensor:
+        return total_variation(img)
+
+
+def total_variation(img: torch.Tensor) -> torch.Tensor:
+    r"""Function that computes Total Variation.
+
+    See :class:`~kornia.losses.TotalVariation` for details.
+    """
+    if not torch.is_tensor(img):
+        raise TypeError(f"Input type is not a torch.Tensor. Got {type(img)}")
+    img_shape = img.shape
+    if len(img_shape) == 3:
+        pixel_dif1 = img[:, 1:, :] - img[:, :-1, :]
+        pixel_dif2 = img[:, :, 1:] - img[:, :, :-1]
+        reduce_axes = (0, 1, 2)
+    elif len(img_shape) == 4:
+        pixel_dif1 = img[:, :, 1:, :] - img[:, :, :-1, :]
+        pixel_dif2 = img[:, :, :, 1:] - img[:, :, :, :-1]
+        reduce_axes = (1, 2, 3)
+    else:
+        raise ValueError("Expected input tensor to be of rank 3 or 4, but got " + str(len(img_shape)))
```

Comment:
I believe, it is not "rank", but ndim. 

"The total number of contravariant and covariant indices of a tensor. The rank R of a tensor is independent of the number of dimensions N of the underlying space."

http://mathworld.wolfram.com/TensorRank.html

Reply:
This is not needed using ellipsis. We can handle (*,H,W)
```python
pixel_dif1 = img[..., 1:, :] - img[..., :-1, :]
pixel_dif2 = img[..., :, 1:] - img[..., :, :-1]
```

Reply:
Besides I believe it can be done with a conv2d using a filter of [1, -1] including padding so that we can recover the full resolution image

Reply:
Whatever is faster is fine to me

Reply:
@edgarriba 
I did think of using convolution. But the example for denoising I have ran in ~9 secs with conv (while with the current diff implementation it ran in ~2 secs). (ran on CPU, not GPU). The time diff was quite surprising to me!

here is the code for total variation implemented in terms of convolution. Let me know if you see any scope of optimizing the conv based implementation and I can try it:
```
class TotalVariation(nn.Module):
    def __init__(self) -> None:
        super(TotalVariation, self).__init__()
        self.conv_3d_1 = torch.nn.Conv2d(in_channels=1, out_channels=1, kernel_size=(1,2), bias=False)
        self.conv_3d_1.weight = torch.nn.Parameter(data=torch.tensor([[[[1.0,-1.0]]]]), requires_grad=False)
        self.conv_3d_2 = torch.nn.Conv2d(in_channels=1, out_channels=1, kernel_size=(2,1), bias=False)
        self.conv_3d_2.weight = torch.nn.Parameter(data=torch.tensor([[[[ 1.0],[-1.0]]]]), requires_grad=False)

    def forward(  # type: ignore
            self, img) -> torch.Tensor:
        #return total_variation(img)
        if not torch.is_tensor(img):
            raise TypeError(f"Input type is not a torch.Tensor. Got {type(img)}")
        img_shape = img.shape
        if not (len(img_shape) == 3 or len(img_shape) == 4):
            raise ValueError("Expected input tensor to be of ndim 3 or 4, but got " + str(len(img_shape)))
        if len(img_shape) == 3:
            # unsqueeze introduces a batch dimension
            img = img.unsqueeze(0)
        num_channels = img_shape[1]
        channel_separated = torch.chunk(img, chunks=num_channels, dim=1)
        # channel_separated is C tensors of size N x 1 x H x W

        to_be_added = []
        for channel in channel_separated:
            to_be_added.append(self.conv_3d_1(channel).abs().sum(dim=(1, 2, 3)))
            to_be_added.append(self.conv_3d_2(channel).abs().sum(dim=(1, 2, 3)))
        total : torch.Tensor = torch.stack(to_be_added).sum(dim=0)
        if len(img_shape) == 3:
            # get rid of the batch dimension
            total = torch.squeeze(total)
        return total


def total_variation(img: torch.Tensor) -> torch.Tensor:
    return TotalVariation()(img)
```

Reply:
filter2D already implements that. Give it a shot and check the performance
https://kornia.readthedocs.io/en/latest/filters.html#kornia.filters.filter2D

Reply:
Tried `kornia.filter.filter2D`, but it is still slow. I think that its just conv in this case is slower than subtraction. I suggest we stick to the current difference based implementation

Note I tried on CPU
```
class TotalVariation(nn.Module):
    def __init__(self) -> None:
        super(TotalVariation, self).__init__()
        self.weight1 = torch.nn.Parameter(data=torch.tensor([[[1.0,-1.0]]]), requires_grad=False)
        self.weight2 = torch.nn.Parameter(data=torch.tensor([[[ 1.0],[-1.0]]]), requires_grad=False)

    def forward(  # type: ignore
            self, img) -> torch.Tensor:
        if not torch.is_tensor(img):
            raise TypeError(f"Input type is not a torch.Tensor. Got {type(img)}")
        img_shape = img.shape
        if not (len(img_shape) == 3 or len(img_shape) == 4):
            raise ValueError("Expected input tensor to be of ndim 3 or 4, but got " + str(len(img_shape)))
        if len(img_shape) == 3:
            # unsqueeze introduces a batch dimension
            img = img.unsqueeze(0)
        num_channels = img_shape[1]
        channel_separated = torch.chunk(img, chunks=num_channels, dim=1)
        # channel_separated is C tensors of size N x 1 x H x W

        to_be_added = []
        for channel in channel_separated:
            to_be_added.append(kornia.filters.filter2D(channel, self.weight1).abs().sum(dim=(1, 2, 3)))

            to_be_added.append(kornia.filters.filter2D(channel, self.weight2).abs().sum(dim=(1, 2, 3)))
        total : torch.Tensor = torch.stack(to_be_added).sum(dim=0)
        if len(img_shape) == 3:
            # get rid of the batch dimension
            total = torch.squeeze(total)
        return total
```

Reply:
Great